server.go

```diff
@@ -111,36 +112,44 @@ func (server *mongoServer) AcquireSocket(poolLimit int, timeout time.Duration) (
 			server.Unlock()
 			return nil, abended, errServerClosed
 		}
-		n := len(server.unusedSockets)
-		if poolLimit > 0 && len(server.liveSockets)-n >= poolLimit {
+
+		if poolLimit > 0 && server.activeSocketsCount > poolLimit {
 			server.Unlock()
 			return nil, false, errPoolLimit
 		}
-		if n > 0 {
-			socket = server.unusedSockets[n-1]
-			server.unusedSockets[n-1] = nil // Help GC.
-			server.unusedSockets = server.unusedSockets[:n-1]
+		unusedSocketsCount := len(server.unusedSockets)
+		if unusedSocketsCount > 0 {
+			socket = server.unusedSockets[unusedSocketsCount-1]
+			server.unusedSockets[unusedSocketsCount-1] = nil // Help GC.
+			server.unusedSockets = server.unusedSockets[:unusedSocketsCount-1]
 			info := server.info
 			server.Unlock()
 			err = socket.InitialAcquire(info, timeout)
 			if err != nil {
 				continue
 			}
 		} else {
+			// try to establish new connection to mongo server
+			server.activeSocketsCount++
 			server.Unlock()
 			socket, err = server.Connect(timeout)
 			if err == nil {
 				server.Lock()
 				// We've waited for the Connect, see if we got
 				// closed in the meantime
 				if server.closed {
+					server.activeSocketsCount--
 					server.Unlock()
 					socket.Release()
 					socket.Close()
 					return nil, abended, errServerClosed
 				}
 				server.liveSockets = append(server.liveSockets, socket)
 				server.Unlock()
+			} else {
+				server.Lock()
+				server.activeSocketsCount--
+				server.Unlock()
```

Comment:
The above 3 diffs could be simplified to a single `server.activeSocketCount++` on line 147.

Reply:
activeSocketCount is used to protect liveSockets list

Reply:
I get the mechanics, my point is that you can simplify all this code by only mutating this variable once. However, not managing this yourself (as described [here](https://github.com/lyft/mgo/pull/6#discussion_r146630373)) would be better.